[aarch64] atan2 intrinsic lowering (p5)

tex3d opened the pull request:
This change is part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

- `VecFuncs.def`: define intrinsic to sleef/armpl mapping
- `LegalizerHelper.cpp`: add missing fewerElementsVector handling for the new atan2 intrinsic
- `AArch64ISelLowering.cpp`: Add arch64 specializations for lowering like neon instructions
- `AArch64LegalizerInfo.cpp`: Legalize atan2.

Part 5 for Implement the atan2 HLSL Function #70096.

tex3d committed:
[aarch64] atan2 intrinsic lowering (p5)

This change is part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

- `VecFuncs.def`: define intrinsic to sleef/armpl mapping
- `LegalizerHelper.cpp`: add missing fewerElementsVector handling for the new atan2 intrinsic
- `AArch64ISelLowering.cpp`: Add arch64 specializations for lowering like neon instructions
- `AArch64LegalizerInfo.cpp`: Legalize atan2.

Part 5 for Implement the atan2 HLSL Function #70096.

tex3d committed:
Move GlobalISel changes from x86 to aarch64

tschuett commented on llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir:
```diff
@@ -0,0 +1,255 @@
+# RUN: llc -verify-machineinstrs -mtriple aarch64--- \
```
Why does this file not start with:
; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
?
Look at other legalize-foo.mir files for prior art in the same directory.

tschuett replied:
```
> ./llvm/utils/update_mir_test_checks.py llvm/test/CodeGen/AArch64/GlobalISel/legalize-atan2.mir
```

tex3d replied:
Before, this test was manually adapted from the pattern used for some of the other arc/hyperbolic intrinsics (such as legalize-atan.mir).  I updated the test to the result of using update_mir_test_checks.py instead.

tschuett replied:
Thanks. Automatism is great and the header gives us confidence that the check lines are correct.

farzonl commented on llvm/test/CodeGen/AArch64/replace-with-veclib-armpl.ll:
```diff
@@ -15,7 +15,7 @@ declare <vscale x 2 x double> @llvm.cos.nxv2f64(<vscale x 2 x double>)
 declare <vscale x 4 x float> @llvm.cos.nxv4f32(<vscale x 4 x float>)
```
I was expecting 
- `llvm/test/CodeGen/AArch64/replace-with-veclib-sleef.ll` 
- `llvm/test/CodeGen/AArch64/replace-with-veclib-sleef-scalable.ll`
- `llvm/test/CodeGen/AArch64/vec-libcalls.ll`
- `llvm/test/Transforms/LoopVectorize/AArch64/veclib-calls-libsystem-darwin.ll`
- `llvm/test/Transforms/LoopVectorize/AArch64/veclib-intrinsic-calls.ll`
I think you may have done some of these in another PR. But want to know what the plan is to do the rest?

tex3d committed:
Add missing veclib tests

tex3d committed:
Use update_mir_test_checks.py to update legalize-atan2.mir

Before, this test was manually adapted from the pattern used for some of the other arc/hyperbolic intrinsics (such as legalize-atan.mir).  This changes the test to the result of using update_mir_test_checks.py instead.

tex3d commented:
Failures in buildkite are unrelated.  I might update the branch just to kick it off again, hopefully with that issue fixed.

tschuett commented:
Known Windows issue:
https://github.com/llvm/llvm-project/issues/112961

tex3d commented:
@tschuett 
> Known Windows issue: #112961

Thanks, I rebased, so hopefully those pipelines will pass now.  If the change looks good to you, would you like to approve?
